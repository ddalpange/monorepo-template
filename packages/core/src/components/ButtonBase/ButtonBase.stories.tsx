import {Story} from '@storybook/react';
import React from 'react';

import ButtonBase from '.';

export const Template: Story = args => {
  return <ButtonBase {...args} />;
};

Template.args = {children: 'Lorem Ipsum'};

export default {
  title: 'Components/ButtonBase',
  component: ButtonBase,
};
