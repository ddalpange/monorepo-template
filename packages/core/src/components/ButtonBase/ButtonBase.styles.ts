import {css} from '@emotion/react';
import styled from '@emotion/styled';

interface ButtonBaseStyleProps {
  isFullWidth?: boolean;
}

export const ButtonBase = styled.button<ButtonBaseStyleProps>`
  display: inline-flex;
  align-items: center;
  padding: 0;
  margin: 0;
  border: 0;
  background: none;
  cursor: pointer;

  :disabled {
    cursor: not-allowed;
  }

  ${({isFullWidth}) =>
    isFullWidth &&
    css`
      width: 100%;
    `}
`;
