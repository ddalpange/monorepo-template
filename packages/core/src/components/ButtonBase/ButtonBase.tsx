import React from 'react';

import * as S from './ButtonBase.styles';

export interface ButtonBaseProps extends Omit<React.ComponentPropsWithoutRef<'button'>, 'disabled'> {
  isFullWidth?: boolean;
  isDisabled?: boolean;
}

const ButtonBase = React.forwardRef<HTMLButtonElement, ButtonBaseProps>(({isFullWidth, isDisabled, ...props}, ref) => {
  return <S.ButtonBase ref={ref} disabled={isDisabled} isFullWidth={isFullWidth} {...props} />;
});

export default React.memo(ButtonBase);

ButtonBase.displayName = 'ButtonBase';
