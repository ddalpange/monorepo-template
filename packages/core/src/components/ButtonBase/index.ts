export {default} from './ButtonBase';
export * from './ButtonBase';
export {ButtonBase as ButtonBaseStyle} from './ButtonBase.styles';
