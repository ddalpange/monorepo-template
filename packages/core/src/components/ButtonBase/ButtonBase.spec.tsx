import {render, screen} from '@testing-library/react';
import React from 'react';

import ButtonBase from '.';

describe('<ButtonBase />', () => {
  it('should render correctly', () => {
    const {container} = render(<ButtonBase>Lorem Ipsum</ButtonBase>);
    expect(container).toMatchSnapshot();
  });

  it('should set width to `100%` when `fullWidth` prop is `true`', () => {
    render(<ButtonBase isFullWidth>Lorem Ipsum</ButtonBase>);

    expect(screen.getByRole('button')).toHaveStyle({
      width: '100%',
    });
  });

  it('should not set width to `100%` when `fullWidth` prop is `false`', () => {
    render(<ButtonBase>Lorem Ipsum</ButtonBase>);

    expect(screen.getByRole('button')).not.toHaveStyle({
      width: '100%',
    });
  });

  it.each<[boolean, boolean]>([
    [false, false],
    [true, true],
  ])('should `disabled` property to `%s` when `isDisabled` property is `%s`', (isDisabled, expected) => {
    const {container} = render(<ButtonBase isDisabled={isDisabled}>Lorem Ipsum</ButtonBase>);

    expect(container.firstChild).toHaveProperty('disabled', expected);
  });

  it('should accept type property', () => {
    render(<ButtonBase type="submit">Lorem Ipsum</ButtonBase>);
    const button = screen.queryByText('Lorem Ipsum');
    expect(button).toBeInTheDocument();
    expect(button).toHaveAttribute('type', 'submit');
  });
});
