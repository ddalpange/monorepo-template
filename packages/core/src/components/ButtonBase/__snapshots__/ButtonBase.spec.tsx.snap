// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonBase /> should render correctly 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0;
  margin: 0;
  border: 0;
  background: none;
  cursor: pointer;
}

.emotion-0:disabled {
  cursor: not-allowed;
}

<div>
  <button
    class="emotion-0"
  >
    Lorem Ipsum
  </button>
</div>
`;
